CancellationException thrown when calling concurrency @Asynchorous method

## Environment Details
* GlassFish Version (and build number): GlassFish 7.0.0 M10
* JDK version: 17
* OS: Windows 10 pro 64bit
* Database: Built-in Derby

----------

## Problem Description

The concurrency @Asynchorous is used to replace the EJB @Asynchorous, so I tried to use concurrency version in a CDI bean.

```java
@ApplicationScoped
public class TodoService {

    @Resource(lookup = "java:module/concurrent/MyExecutor")
    ManagedExecutorService executorService;

    @PersistenceContext
    EntityManager entityManager;

    @Asynchronous
    public CompletableFuture<List<Todo>> getAllTodosAsync() {
        Supplier<List<Todo>> todos = () -> entityManager.createQuery("select t from Todo t", Todo.class).getResultList();
        return executorService.supplyAsync(todos);
    }
...
```

And I wrote the arquillian tests.

```java
@Test
@Order(3)
public void testGetAllTodosAsync() throws Exception {
    todoService.getAllTodosAsync()
            .thenAccept(
                    todos -> {
                        LOGGER.log(Level.INFO, "getAllTodosAsync result: {0}", new Object[]{todos});
                        assertNotNull(todos);
                        assertEquals(3, todos.size());
                    }
            )
            .join();

}
```

## Steps to reproduce

1. Get a copy of my example codes: https://github.com/hantsy/jakartaee10-sandbox/tree/master/concurrency
2. run `mvn clean verify -Parq-glassfish-managed -D"it.test=TodoServiceTest"`
3. It will throw the following exception.
  ```bash
Error:  com.example.it.TodoServiceTest.testGetAllTodosAsync  Time elapsed: 0.067 s  <<< ERROR!
java.util.concurrent.CompletionException: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2480)
	at org.glassfish.enterprise.concurrent.AsynchronousInterceptor.lambda$intercept$0(AsynchronousInterceptor.java:78)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.glassfish.enterprise.concurrent.internal.ManagedFutureTask.run(ManagedFutureTask.java:119)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
	at org.glassfish.enterprise.concurrent.ManagedThreadFactoryImpl$ManagedThread.run(ManagedThreadFactoryImpl.java:322)
Caused by: java.util.concurrent.CancellationException
	at java.base/java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2478)
	... 8 more
  ```

In the GlassFish server.log, I notice there is a message like this.

```bash
@Asynchronous method does not call Asynchronous.Result.complete()...
```

But the EJB @Asynchronous works well.

```java
@Stateless
public class EjbTodoService {
    @PersistenceContext(unitName = "defaultPU")
    EntityManager entityManager;

    @Resource
    ManagedExecutorService executorService;

    @jakarta.ejb.Asynchronous
    public Future<List<Todo>> getAllTodosEjbAsync() {
        Callable<List<Todo>> callable = () -> entityManager.createQuery("select t from Todo t", Todo.class).getResultList();
        return executorService.submit(callable);
    }

}
````

```java
@Test
@Order(1)
public void testGetAllTodosEjbAsync() throws Exception {
    var todos = ejbTodoService.getAllTodosEjbAsync().get(1000, TimeUnit.MILLISECONDS);

    LOGGER.log(Level.INFO, "getAllTodosAsync result: {0}", new Object[]{todos});
    assertNotNull(todos);
    assertEquals(3, todos.size());
}
```

--- a/src/main/java/org/glassfish/concurro/cdi/asynchronous/AsynchronousInterceptor.java
+++ b/src/main/java/org/glassfish/concurro/cdi/asynchronous/AsynchronousInterceptor.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022, 2025 Contributors to the Eclipse Foundation.
+ * Copyright (c) 2022, 2025, 2026 Contributors to the Eclipse Foundation.
  * Copyright (c) 2024 Payara Foundation and/or its affiliates.
  *
  * This program and the accompanying materials are made available under the
@@ -118,16 +118,16 @@ public class AsynchronousInterceptor {
                     method);
                 resultFuture.cancel(true);
             }
-            return;
+        } else {
+            // The method returned a different future; complete resultFuture when it completes.
+            returnedFuture.whenComplete((value, throwable) -> {
+                if (throwable != null) {
+                    resultFuture.completeExceptionally(throwable);
+                } else {
+                    resultFuture.complete(value);
+                }
+            });
         }
-        // The method returned a different future; complete resultFuture when it completes.
-        returnedFuture.whenComplete((value, throwable) -> {
-            if (throwable != null) {
-                resultFuture.completeExceptionally(throwable);
-            } else {
-                resultFuture.complete(value);
-            }
-        });
     }
 
     private static <T> T lookupMES(Class<T> cls, String executor, String methodName) throws RejectedExecutionException {

--- a/src/main/java/org/glassfish/concurro/cdi/asynchronous/AsynchronousInterceptor.java
+++ b/src/main/java/org/glassfish/concurro/cdi/asynchronous/AsynchronousInterceptor.java
@@ -31,7 +31,6 @@ import java.lang.reflect.Method;
 import java.time.ZoneId;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ExecutionException;
 import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.ScheduledFuture;
 import java.util.function.Consumer;
@@ -76,33 +75,59 @@ public class AsynchronousInterceptor {
         CompletableFuture<Object> resultFuture = new ManagedCompletableFuture<>(mes);
         mes.submit(() -> {
             Asynchronous.Result.setFuture(resultFuture);
-            CompletableFuture<Object> returnedFuture = resultFuture;
             try {
-                // the asynchronous method is responsible for calling Asynchronous.Result.complete()
-                returnedFuture = (CompletableFuture<Object>) context.proceed();
+                // The asynchronous method either completes Asynchronous.Result and returns it,
+                // or returns a different CompletableFuture instance.
+                CompletableFuture<Object> returnedFuture = (CompletableFuture<Object>) context.proceed();
+                completeResultFuture(method, resultFuture, returnedFuture);
             } catch (Exception ex) {
                 resultFuture.completeExceptionally(ex);
             } finally {
-                // Check if Asynchronous.Result is not completed?
-                if (!returnedFuture.isDone()) {
-                    LOG.log(ERROR,
-                        "Method annotated with @Asynchronous did not call Asynchronous.Result.complete() at its end: {0}",
-                        method);
-                    Asynchronous.Result.getFuture().cancel(true);
-                }
-                if (returnedFuture != Asynchronous.Result.getFuture()) {
-                    // if the asynchronous methods returns a different future, use this to complete the resultFuture
-                    try {
-                        resultFuture.complete(returnedFuture.get());
-                    } catch (InterruptedException | ExecutionException e) {
-                        resultFuture.completeExceptionally(e);
-                    }
-                }
                 // cleanup after asynchronous call
                 Asynchronous.Result.setFuture(null);
             }
         });
         return resultFuture;
+    }
+
+    /**
+     * Bridges the future returned by an {@code @Asynchronous} method to the {@code resultFuture}
+     * handed to the caller.
+     * <p>
+     * Two cases are supported per the Jakarta Concurrency specification:
+     * <ul>
+     * <li>The method completes {@link Asynchronous.Result} and returns it (the same instance as
+     * {@code resultFuture}). In that case the result future is already completed; if it is not, the
+     * method forgot to call {@code Asynchronous.Result.complete()} and the result future is
+     * cancelled.</li>
+     * <li>The method returns a <em>different</em> {@link CompletableFuture}. Its completion (value or
+     * exception) is propagated to {@code resultFuture} without blocking the executor thread.</li>
+     * </ul>
+     *
+     * @param method the intercepted method, used for diagnostics only
+     * @param resultFuture the future returned to the caller (also bound to {@link Asynchronous.Result})
+     * @param returnedFuture the future returned by the method body
+     */
+    static void completeResultFuture(Method method, CompletableFuture<Object> resultFuture,
+            CompletableFuture<Object> returnedFuture) {
+        if (returnedFuture == resultFuture) {
+            // The method was expected to complete Asynchronous.Result itself.
+            if (!returnedFuture.isDone()) {
+                LOG.log(ERROR,
+                    "Method annotated with @Asynchronous did not call Asynchronous.Result.complete() at its end: {0}",
+                    method);
+                resultFuture.cancel(true);
+            }
+            return;
+        }
+        // The method returned a different future; complete resultFuture when it completes.
+        returnedFuture.whenComplete((value, throwable) -> {
+            if (throwable != null) {
+                resultFuture.completeExceptionally(throwable);
+            } else {
+                resultFuture.complete(value);
+            }
+        });
     }
 
     private static <T> T lookupMES(Class<T> cls, String executor, String methodName) throws RejectedExecutionException {

--- a/src/test/java/org/glassfish/concurro/cdi/asynchronous/AsynchronousInterceptorTest.java
+++ b/src/test/java/org/glassfish/concurro/cdi/asynchronous/AsynchronousInterceptorTest.java
@@ -16,9 +16,94 @@
 
 package org.glassfish.concurro.cdi.asynchronous;
 
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
 import org.junit.jupiter.api.Test;
 
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
 public class AsynchronousInterceptorTest {
+
+    /**
+     * Regression test for <a href="https://github.com/eclipse-ee4j/glassfish/issues/24203">#24203</a>:
+     * when an {@code @Asynchronous} method returns a <em>different</em>, not-yet-completed
+     * CompletableFuture, the result future handed to the caller must not be cancelled. It must be
+     * completed with the same result once the returned future completes.
+     */
+    @Test
+    public void differentNotYetCompletedFutureIsBridgedNotCancelled() throws Exception {
+        CompletableFuture<Object> resultFuture = new CompletableFuture<>();
+        CompletableFuture<Object> returnedFuture = new CompletableFuture<>();
+
+        AsynchronousInterceptor.completeResultFuture(anyMethod(), resultFuture, returnedFuture);
+
+        // The returned future is still running: the result future must stay pending, not be cancelled.
+        assertFalse(resultFuture.isDone(), "result future must not be completed yet");
+        assertFalse(resultFuture.isCancelled(), "result future must not be cancelled");
+
+        returnedFuture.complete("hello");
+
+        assertEquals("hello", resultFuture.get(5, SECONDS));
+        assertFalse(resultFuture.isCancelled());
+    }
+
+    @Test
+    public void differentFutureExceptionIsPropagated() {
+        CompletableFuture<Object> resultFuture = new CompletableFuture<>();
+        CompletableFuture<Object> returnedFuture = new CompletableFuture<>();
+
+        AsynchronousInterceptor.completeResultFuture(anyMethod(), resultFuture, returnedFuture);
+
+        IllegalStateException failure = new IllegalStateException("boom");
+        returnedFuture.completeExceptionally(failure);
+
+        ExecutionException thrown = assertThrows(ExecutionException.class, resultFuture::get);
+        assertSame(failure, thrown.getCause());
+        assertFalse(resultFuture.isCancelled());
+    }
+
+    @Test
+    public void alreadyCompletedDifferentFutureCompletesResult() throws Exception {
+        CompletableFuture<Object> resultFuture = new CompletableFuture<>();
+        CompletableFuture<Object> returnedFuture = CompletableFuture.completedFuture("done");
+
+        AsynchronousInterceptor.completeResultFuture(anyMethod(), resultFuture, returnedFuture);
+
+        assertEquals("done", resultFuture.get(5, SECONDS));
+    }
+
+    @Test
+    public void sameFutureCompletedByMethodIsLeftIntact() throws Exception {
+        CompletableFuture<Object> resultFuture = new CompletableFuture<>();
+        resultFuture.complete("via Asynchronous.Result");
+
+        AsynchronousInterceptor.completeResultFuture(anyMethod(), resultFuture, resultFuture);
+
+        assertEquals("via Asynchronous.Result", resultFuture.get(5, SECONDS));
+        assertFalse(resultFuture.isCancelled());
+    }
+
+    @Test
+    public void sameFutureNotCompletedByMethodIsCancelled() {
+        CompletableFuture<Object> resultFuture = new CompletableFuture<>();
+
+        AsynchronousInterceptor.completeResultFuture(anyMethod(), resultFuture, resultFuture);
+
+        // The method forgot to call Asynchronous.Result.complete(): the result future is cancelled.
+        assertTrue(resultFuture.isCancelled());
+    }
+
+    private static java.lang.reflect.Method anyMethod() {
+        // Any method works: it is only used for diagnostic logging.
+        return AsynchronousInterceptorTest.class.getDeclaredMethods()[0];
+    }
+
     @Test
     public void testGettingCronTriggerFromScheduleWithCronExpression() {
         var scheduleWithCronExpression = ScheduleStub.newScheduleWithCronExpression("*/5 * * * * *");

